galley/pkg/config/source/kube/inmemory/kubesource.go

```diff
@@ -303,10 +304,69 @@ func (s *KubeSource) parseChunk(r *collection.Schemas, name string, lineNum int,
 		return kubeResource{}, err
 	}
 
+	// Build flat map for analyzers if the line JSON object exists, if the YAML text is ill-formed, this will be nil
+	fieldMap := make(map[string]int)
+
+	// yamlv3.Node contains information like line number of the node, which will be used with its name to construct the field map
+	yamlChunkNode := yamlv3.Node{}
+	err = yamlv3.Unmarshal(yamlChunk, &yamlChunkNode)
+	if err == nil && len(yamlChunkNode.Content) == 1 {
+
+		// Get the Node that contains all the YAML chunk information
+		yamlNode := yamlChunkNode.Content[0]
+
+		BuildFieldPathMap(yamlNode, lineNum, "", fieldMap)
+	}
+
 	pos := rt.Position{Filename: name, Line: lineNum}
 	return kubeResource{
 		schema:   schema,
 		sha:      sha1.Sum(yamlChunk),
-		resource: rt.ToResource(objMeta, schema, item, &pos),
+		resource: rt.ToResource(objMeta, schema, item, &pos, fieldMap),
 	}, nil
 }
+
+// BuildFieldPathMap builds the flat map for each field of the YAML resource
+func BuildFieldPathMap(yamlElementNode *yamlv3.Node, startLineNum int, curPath string, fieldPathMap map[string]int) {
+	// If no content in the node, terminate the DFS search
+	if len(yamlElementNode.Content) == 0 {
+		return
+	}
+
+	i := 0
+	NodeContent := yamlElementNode.Content
+	// Iterate content by a step of 2, because in the content array the value is in the key's next index position
+	for i < len(NodeContent)-1 {
+		// Two condition, i + 1 positions have no content, which means they have the format like "key: value", then build the map
+		// Or i + 1 has contents, which means "key:\n  value...", then perform one more DFS search
+		keyNode := NodeContent[i]
+		valueNode := NodeContent[i+1]
+		pathKeyForMap := curPath + "." + keyNode.Value
+		if len(valueNode.Content) == 0 {
+			// minus one because startLineNum starts at line 1, and yamlv3.Node.line also starts at line 1
+			fieldPathMap["{"+pathKeyForMap+"}"] = valueNode.Line + startLineNum - 1
+		} else {
+			// Next iteration for child node
+			if valueNode.Tag == "!!seq" {
+				j := 0
+				for j < len(valueNode.Content) {
+					pathWithIndex := pathKeyForMap + "[" + fmt.Sprintf("%d", j) + "]"
+
+					// Array with values or array with maps
+					if len(valueNode.Content[j].Content) == 0 {
+						fieldPathMap["{"+pathWithIndex+"}"] = valueNode.Content[j].Line + startLineNum - 1
+					} else {
+						BuildFieldPathMap(valueNode.Content[j], startLineNum, pathWithIndex, fieldPathMap)
+					}
+
+					j++
+				}
+			} else if valueNode.Tag == "!!map" {
+				BuildFieldPathMap(valueNode, startLineNum, pathKeyForMap, fieldPathMap)
+			}
+		}
```

Comment:
Can you flatten this to just be if len(valueNode.Content) == 0 {} else if valueNode.Tag == "!!seq" {} else if valueNode.Tag == "!!map" {}?
If so, please use a switch statement: see https://golang.org/doc/effective_go.html#switch

Reply:
Thanks!